frontend/app/model/MembershipCatalog.scala

```diff
@@ -0,0 +1,219 @@
+
+    def currenciesMismatchErr(tierDetails: PaidTierDetails): NonEmptyList[Err] = {
+      val msg = "The supported currencies for the different plans do not match"
+      NonEmptyList(
+        tierDetails.monthlyPlanDetails.plan -> msg,
+        tierDetails.yearlyPlanDetails.plan -> msg
+      )
+    }
+
+    def currenciesMatch(tierDetails: PaidTierDetails): Boolean =
+      tierDetails.monthlyPlanDetails.currencies == tierDetails.yearlyPlanDetails.currencies
+
+    object PublicPlans {
+      val ids = ratePlanIds
+      val friend = freeTierPlanDetails(FriendTierPlan.public, ids.friend)
+      val staff = freeTierPlanDetails(StaffPlan, ids.staff)
+      val supporterM = paidTierPlanDetails(PaidTierPlan.monthly(Supporter, Public), ids.supporterMonthly)
+      val supporterY = paidTierPlanDetails(PaidTierPlan.yearly(Supporter, Public), ids.supporterYearly)
+      val partnerM = paidTierPlanDetails(PaidTierPlan.monthly(Partner, Public), ids.partnerMonthly)
+      val partnerY = paidTierPlanDetails(PaidTierPlan.yearly(Partner, Public), ids.partnerYearly)
+      val patronM = paidTierPlanDetails(PaidTierPlan.monthly(Patron, Public), ids.patronMonthly)
+      val patronY = paidTierPlanDetails(PaidTierPlan.yearly(Patron, Public), ids.patronYearly)
+      val all = List(friend, staff, supporterM, supporterY, partnerM, partnerY, patronM, patronY)
+    }
+
+    object LegacyPlans {
+      val ids = ratePlanIds.legacy
+      val friend = freeTierPlanDetails(FriendTierPlan.legacy, ids.friend)
+      val supporterM = paidTierPlanDetails(PaidTierPlan.monthly(Supporter, Legacy), ids.supporterMonthly)
+      val supporterY = paidTierPlanDetails(PaidTierPlan.yearly(Supporter, Legacy), ids.supporterYearly)
+      val partnerM = paidTierPlanDetails(PaidTierPlan.monthly(Partner, Legacy), ids.partnerMonthly)
+      val partnerY = paidTierPlanDetails(PaidTierPlan.yearly(Partner, Legacy), ids.partnerYearly)
+      val patronM = paidTierPlanDetails(PaidTierPlan.monthly(Patron, Legacy), ids.patronMonthly)
+      val patronY = paidTierPlanDetails(PaidTierPlan.yearly(Patron, Legacy), ids.patronYearly)
+      val all = List(friend, supporterM, supporterY, partnerM, partnerY, patronM, patronY)
+    }
+
+    val tierPlanDetailsMap: Val[Map[ProductRatePlanId, TierPlanDetails]] =
+      //The failed public plans are discarded as they are going to be accumulated while the tier details are computed
+      Applicative[Val].sequence(PublicPlans.all.filter(_.isSuccess) ++ LegacyPlans.all)
+        .map(_.map { d => d.productRatePlanId -> d }.toMap)
+
+    def paidTierDetails(monthly: Val[PaidTierPlanDetails], yearly: Val[PaidTierPlanDetails]): Val[PaidTierDetails] =
+      (monthly |@| yearly)(PaidTierDetails).assert(currenciesMismatchErr)(currenciesMatch)
+
+    object Details {
+      private val P = PublicPlans
+      val friend = P.friend.map(FreeTierDetails)
+      val staff = P.staff.map(FreeTierDetails)
+      val supporter = paidTierDetails(P.supporterM, P.supporterY)
```

Comment:
You shouldn't have to say `paidTierDetails()` three times! Would really prefer fewer declared variables, better to loop over lists of tiers.